README.md
# NAME

Navet::ePersondata - Simple framework for the Navet ePersondata web services

# SYNOPSIS

      use Navet::ePersondata::Personpost;

      my $ep = Navet::ePersondata::Personpost->new(    
          # Set proxy to test service instead of production 
          soap_options => {
              proxy => 'https://www2.test.skatteverket.se/na/na_epersondata/V2/personpostXML'
          },
          pkcs12_file => '/path/to/certificate.p12',
          pkcs12_password => '5085873593180405',
          OrgNr => '162021004748',
          BestallningsId => '00000079-FO01-0001',
      );

      my $node = $ep->find_first({PersonId => '198602212394'})
          or die "Failed: ", $ep->error->{message};
      
      my ($namn_node) = $node->findnodes('./Personpost/Namn');
      print "The name is ", join ' ', $namn_node->findvalue('./Fornamn'),
          $namn_node->findvalue('./Efternamn'), "\n";
      
      # Using Namsokning and termdata format:
      
      use Navet::ePersondata::Namnsokning;
      
      my $ep = Navet::ePersondata::Namnsokning->new(
          pkcs12_file => '/path/to/certificate.p12',
          pkcs12_password => '5085873593180405',
          OrgNr => '162021004748',
          BestallningsId => '00000079-FO01-0001',
          format => 'termdata'
      );
      
      my $query = {
          Kon => 'K',
          Fornamn => 'sar*',
          Postort => 'Boden'            
      };
      
      foreach my $termdata ( $ep->find_all($query) ) {
          my ($first_name, $middle_name, $last_name) =
              ($termdata->{'01012'}, $termdata->{'01013'}, $termdata->{'01014'});
              
          $full_name = $first_name
              . ($middle_name ? " $middle_name" : '') . " $last_name";
          print "$full_name\n";
      }
    

# DESCRIPTION

This package is the base class for talking with  Web Services - ePersondata
provided by skatteverket.se. The specific modules to use are
[Navet::ePersondata::PersonPost](https://metacpan.org/pod/Navet::ePersondata::PersonPost) and  [Navet::ePersondata:Namsokning](https://metacpan.org/pod/Navet::ePersondata:Namsokning) corresponding
to Web Service Personpost and Namnsökning. Using these services requires agreement
and access granted by Skatteverket in Sweden and also a special certificate. 

# METHODS

## new( %options )

    my $personpost_client = Navet::ePersondata::Personpost->new(
       pkcs12_file => '/path/to/certificate_file',
       pkcs12_password  => 'secret', 
       OrgNr => '162021004748',  
       BestallningsId => '00000079-FO01-0001',
       format => 'termdata', # ('termdata' or 'xml') default 'xml'
       soap_options => { proxy => 'https..', ...}  
    );

This constructor requires `pkcs12_file`, `pkcs12_password`, `OrgNr` and
`BestallningsId` or it will croak.

See information from Skatteverket in how to optain these. See
[SOAP::XML::Client](https://metacpan.org/pod/SOAP::XML::Client) for `soap_options` if they are needed.

## find\_all( \\%query\_data )

    my @hits = $personpost_client->find_all( {PersonId => '196601010101'} )
    

query\_data should contain the query parameters to the service.

Performs a SOAP call to the service and returns a list of objects that match the query.
The format of the objects is determined by what format was specified in the constructor:
[XML::LibXML::Node](https://metacpan.org/pod/XML::LibXML::Node) if it was 'xml', references to a flat hash of termcodes, termdata
if it was 'termdata'. Returns empty list if there was an error or if no record was found.

## find\_first( \\%query\_data )

Like find\_all except returns a single entry (first found) instead of a list.
If not found or an error occurred it returns `undef`.

## error()

    if (my $err = $client->error) {
        die "Error: message=", $err->{message}, "\n";
    }

Returns a reference to a hash of the error information if the last call resulted in an error.
Returns `undef` if there was no error. The hash can contain the following fields:

    message          - error text
    soap_faultcode   - SOAP faultcode from /Envelope/Body/Fault/faultscode
    soap_faultstring - SOAP faultstring from /Envelope/BodyFault/faultstring
    sv_Felkod        - Extra error code provided by Skatteverket
    sv_Beskrivning   - Extra description provided by Skatteverket
    raw_error        - Unparsed error text (can be XML, HTML or plain text)
    https_status     - HTTP status code

Only `message` and `raw_error` are garanteed to be set.

# SEE ALSO

[Navet::ePersondata::Personpost](https://metacpan.org/pod/Navet::ePersondata::Personpost)

[Navet::ePersondata::Namnsokning](https://metacpan.org/pod/Navet::ePersondata::Namnsokning)

[SOAP::XML::Client](https://metacpan.org/pod/SOAP::XML::Client)

[https://skatteverket.se/foretagochorganisationer/myndigheter/informationsutbytemellanmyndigheter/navethamtauppgifteromfolkbokforing.html](https://skatteverket.se/foretagochorganisationer/myndigheter/informationsutbytemellanmyndigheter/navethamtauppgifteromfolkbokforing.html)  

# AUTHOR

Snorri Briem <snorri.briem@ub.lu.se>

# COPYRIGHT

Copyright 2018- Lund University Library

# LICENSE

This is free software; you can redistribute it and/or modify it under the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Navet-ePersondata

0.02
  - Updated Navet test URLs and test certificate
0.01
  - Initial release


lib/Navet/ePersondata/Personpost.pm
package Navet::ePersondata::Personpost;

use strict;
use warnings;

use Navet::ePersondata;
use Moo;

extends 'Navet::ePersondata';

sub _args_xml {
     my ($self, $arg) = @_;

     my $pnr = $self->_escape_string( $arg->{PersonId} );
     $self->_bestallning_xml .  "<PersonId>$pnr</PersonId>";
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Navet::ePersondata::Personpost - Talks to the Navet Personpost web service

=head1 SYNOPSIS

    use Navet::ePersondata::Personpost;

    my $ep = Navet::ePersondata::Personpost->new(    
        # Set proxy to test service instead of production 
        soap_options => {
            proxy => 'https://www2.test.skatteverket.se/na/na_epersondata/V2/personpostXML'
        },
        pkcs12_file => '/path/to/certificate.p12',
        pkcs12_password => '5085873593180405',
        OrgNr => '162021004748',
        BestallningsId => '00000079-FO01-0001',
    );

    my $node = $ep->find_first({PersonId => '198602212394'})
        or die "Failed: ", $ep->error->{message};
    
    my ($namn_node) = $node->findnodes('./Personpost/Namn');
    print "The name is ", join ' ', $namn_node->findvalue('./Fornamn'),
        $namn_node->findvalue('./Efternamn'), "\n";


=head1 DESCRIPTION

Navet::ePersondata::Personpost is class for working with the Navet ePersondata
Personpost web service provided by skatteverket.se.

=head1 METHODS

=head2 new( %options )

=head2 find_all( \%query_data )

=head2 find_first( \%query_data )
   
=head2 error()

This class inherits the above methods from L<Navet::ePersondata>. The
C<query_data> parameter should contain C<PersonId>, which stands for a swedish
Personnummer/Samordningsnummer) when using the C<find_all> or C<find_one>
methods. Example:

    $personpost->find_first( {PersonId => '198602212394'} )

=head1 SEE ALSO

L<Navet::ePersondata>

L<Navet::ePersondata::Namnsokning>

L<SOAP::XML::Client>

L<https://skatteverket.se/foretagochorganisationer/myndigheter/informationsutbytemellanmyndigheter/navethamtauppgifteromfolkbokforing.html>  

=head1 AUTHOR

Snorri Briem E<lt>snorri.briem@ub.lu.seE<gt>

=head1 COPYRIGHT

Copyright 2018- Lund University Library

=head1 LICENSE

This is free software; you can redistribute it and/or modify it under the same terms as the Perl 5 programming language system itself.

=cut


lib/Navet/ePersondata/Namnsokning.pm
package Navet::ePersondata::Namnsokning;

use strict;
use warnings;

use Moo;
use Navet::ePersondata;

our $VERSION = '0.02';

extends 'Navet::ePersondata';

my @REQUEST_ELEMENTS = qw(
Adress
EfterMellanNamn
FodelsetidFrom
FodelsetidTom
Fornamn
Kon
PostnummerFrom
PostnummerTom
Postort
Kategori
);


sub _args_xml {
    my ($self, $arg) = @_;

    my $xml =
        $self->_bestallning_xml . 
        '<SokvillkorNamn xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">' .
            join ('', map {
                my $val = $self->_escape_string($arg->{$_} // '');
                my $nil = length $val
                        ? ''
                        : ' xsi:nil="true"';  
                "<$_$nil>$val</$_>"
            } @REQUEST_ELEMENTS) .
            '</SokvillkorNamn>';
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Navet::ePersondata::Namsokning - Talks to the Namsokning web service that is provided by skattaverket.se

=head1 SYNOPSIS

    use Navet::ePersondata::Namnsokning;
    
    my $ep = Navet::ePersondata::Namnsokning->new(
        pkcs12_file => '/path/to/certificate.p12',
        pkcs12_password => '5085873593180405',
        OrgNr => '162021004748',
        BestallningsId => '00000079-FO01-0001',
    );
    
    my $query = {
        Kon => 'K',
        Fornamn => 'sar*',
        Postort => 'Boden'            
    };
    
    my ($namn_node) = $node->findnodes('/Personpost/Namn');
    print "The name is ", join ' ', $namn_node->findvalue('./Fornamn'),
        $namn_node->findvalue('./Efternamn'), "\n";

=head1 DESCRIPTION

Navet::ePersondata::Namnsokning is class for working with the Navet ePersondata
Namnsökning web service provided by skatteverket.se.

=head1 METHODS

=head2 new( %options )

=head2 find_all( \%query_data )

=head2 find_first( \%query_data )
   
=head2 error()

This class inherits the above methods from L<Navet::ePersondata>. The
C<query_data> parameter should contain a combination (with some restrictions) of the following search
fields: C<Adress>, C<EfterMellanNamn>, C<FodelsetidFrom>, C<FodelsetidTom>,
C<Fornamn>, C<Kon>, C<PostnummerFrom>, C<PostnummerTom>, C<Postort>,
C<Kategori>. Example:

    my @hits = $ep->get_data({
        Kon => 'K',
        Fornamn => 'sar*',
        Postort => 'Boden'
    });

=head1 SEE ALSO

L<Navet::ePersondata>

L<Navet::ePersondata::Personpost>

L<SOAP::XML::Client>

L<https://skatteverket.se/foretagochorganisationer/myndigheter/informationsutbytemellanmyndigheter/navethamtauppgifteromfolkbokforing.html>  

=head1 AUTHOR

Snorri Briem E<lt>snorri.briem@ub.lu.seE<gt>

=head1 COPYRIGHT

Copyright 2018- Lund University Library

=head1 LICENSE

This is free software; you can redistribute it and/or modify it under the same terms as the Perl 5 programming language system itself.

=cut


lib/Navet/ePersondata.pm
package Navet::ePersondata;

use strict;
use warnings;

use Carp;
use SOAP::XML::Client::Generic;
use Crypt::SSLeay;
use Net::SSL;
use XML::LibXML;
use Moo;

our $VERSION = '0.02';

has OrgNr => ( is => 'ro', required => 1);
has BestallningsId => (is => 'ro', required => 1);
has pkcs12_file => ( is => 'ro', required => 1);
has pkcs12_password => ( is => 'ro', required => 1);

has format => (
    is => 'ro',
    default => sub {'xml'},
    isa => sub { croak "$_[0] is not a valid format" unless $_[0] =~ m/^(?:xml|termdata)$/ } 
);
has soap_options => ( is => 'ro', default => sub {{}});
has soap => (is  => 'ro', lazy => 1, builder => 1);
has error => ( is => 'ro', clearer => 1, lazy =>1, builder =>1);

my %SOAP_DEFAULT_OPTIONS = (
    uri   => 'http://xmls.skatteverket.se/se/skatteverket/folkbokforing/na/epersondata/V1',
    xmlns => 'http://xmls.skatteverket.se/se/skatteverket/folkbokforing/na/epersondata/V1',
    strip_default_xmlns => 1,
);

sub BUILD {
    my ($self, $args) = @_;
    
    if ( my $proxy = $self->soap_options->{proxy} ) {
        if ( ($self->format eq 'xml' && $proxy !~ m/XML$/) ||
             ($self->format eq 'termdata' && $proxy =~ m/XML$/)
           ) {
            croak "Incompatible values for format and soap proxy";
        }
    }
}


sub _build_soap {
    my $self = shift;

    my %arg = %SOAP_DEFAULT_OPTIONS;
    
    my $xml = $self->format  eq 'xml' ? 'XML' : '';
    my $name = $self->_service_name;
    $arg{proxy} = "https://www2.skatteverket.se/na/na_epersondata/V2/$name$xml";

    SOAP::XML::Client::Generic->new({
        %arg,
        %{$self->soap_options},
      });  
}


sub _service_name  {
    my $self = shift;
    my $name = ref $self;
    $name =~ s/^.+:://;
    return "\l$name";
}

sub _set_ssl_env {
    my $self = shift;
    
    $ENV{HTTPS_PKCS12_FILE}     =  $self->pkcs12_file; 
    $ENV{HTTPS_PKCS12_PASSWORD} =  $self->pkcs12_password;
    $ENV{PERL_LWP_SSL_VERIFY_HOSTNAME} = 0;
    $ENV{PERL_NET_HTTPS_SSL_SOCKET_CLASS}= 'Net::SSL';
}

sub _escape_string {
    my ($self, $text) = @_;
    $text =~ s/&/&amp;/go;
    $text =~ s/</&lt;/go;
    $text =~ s/>/&gt;/go;
    $text =~ s/'/&apos;/go;
    $text =~ s/"/&quot;/go;
    return $text;
}


sub _bestallning_xml {
    my ($self) = @_;
    my $BestallningsId = $self->_escape_string($self->BestallningsId);
    my $OrgNr = $self->_escape_string($self->OrgNr);
"<Bestallning>
<OrgNr>$OrgNr</OrgNr><BestallningsId>$BestallningsId</BestallningsId>
</Bestallning>
";
}

sub _find_records {
    my ($self, $arg, $xpath) = @_;

    my $soap = $self->soap;
    $self->_set_ssl_env();
    
    $self->clear_error();
    
    if ( $soap->fetch({
        method => (ucfirst($self->_service_name) . "Request"),
        xml => ($self->_bestallning_xml . $self->_args_xml($arg)),
    }) ) {
        return $soap->results_xml->findnodes($xpath);
    }

    return; #failed     
}



sub _parse_termdata {
    my ($self, $data, $only_first) = @_;
    my @records;
    my $temp_record;
    
    
    for (split /\n/, $data) {
        my ($a, $c, $d) = split / /, $_, 3;
        next unless $a;
    
        if ($a eq '#POST_START') {
            $temp_record={};
        }
    
        if ($a eq '#UP') {
            $temp_record->{$c}=$d;
        }
    
        if ($a eq '#POST_SLUT') {
            push @records, $temp_record;
            last if $only_first
        }       
    }

    return \@records;
    
}

sub _find_records_termdata {
    my ($self, $arg, $only_first) = @_;
    
    my $response_ele = ucfirst($self->_service_name) . 'Response';
    my ($result) = $self->_find_records($arg, "//*\[local-name()='$response_ele'\]");
    return unless $result;
    return @{$self->_parse_termdata($result->to_literal, $only_first)}; #check
}



# Public Methods. --------------------------------------------------------------

sub find_all {
    my ($self, $arg) = @_;

    if ($self->format eq 'xml' ) {    
        return $self->_find_records($arg, '//Folkbokforingspost');
    }
    
    #format = termdata
    return $self->_find_records_termdata($arg);
}


sub find_first {
    my ($self, $arg) = @_;

    my $found;
    
    if ($self->format eq 'xml') {
        ($found) = $self->_find_records($arg, '//Folkbokforingspost[1]');
        return $found;
    } else {
        #format = termdata
        ($found) = $self->_find_records_termdata($arg,1);
    }
    return $found;
}


sub _build_error {
    my $self = shift;
    
    my $soap_error = $self->soap->error
        or return;

    my $http_status_line = $self->soap->status
        or return {
            message => $soap_error,
            raw_error => $soap_error,
        };
    
    my $error = { raw_error => $soap_error};
    ($error->{http_status},$error->{message}) = split / /, $http_status_line, 2;
    
    $error->{message} ||= $error->{http_status};
    
    my $edom;
    
    eval {
        $edom = XML::LibXML->load_xml(
            string => \$soap_error,
        );
    };
    
    if ($edom) {
        if (my ($fnode) = $edom->documentElement->findnodes('/S:Envelope/S:Body/S:Fault')) {
            my $faultcode = $fnode->findvalue('./faultcode'),
            my $faultstring = $fnode->findvalue('./faultstring'),
            
            my $Felkod = $fnode->findvalue("//*\[local-name()='Felkod'\]");
            my $Beskrivning = $fnode->findvalue("//*\[local-name()='Beskrivning'\]");
            
            $error->{message} = $Beskrivning || $faultstring;
            $error->{soap_faultcode} = $faultcode if $faultcode;
            $error->{soap_faultstring} = $faultstring if $faultstring;
            $error->{sv_Felkod} = $Felkod if $Felkod;
            $error->{sv_Beskrivning} = $Beskrivning if $Beskrivning;
        }
    } else {
        
    }
    return $error;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Navet::ePersondata - Simple framework for the Navet ePersondata web services

=head1 SYNOPSIS

    use Navet::ePersondata::Personpost;

    my $ep = Navet::ePersondata::Personpost->new(    
        # Set proxy to test service instead of production 
        soap_options => {
            proxy => 'https://www2.test.skatteverket.se/na/na_epersondata/V2/personpostXML'
        },
        pkcs12_file => '/path/to/certificate.p12',
        pkcs12_password => '5085873593180405',
        OrgNr => '162021004748',
        BestallningsId => '00000079-FO01-0001',
    );

    my $node = $ep->find_first({PersonId => '198602212394'})
        or die "Failed: ", $ep->error->{message};
    
    my ($namn_node) = $node->findnodes('./Personpost/Namn');
    print "The name is ", join ' ', $namn_node->findvalue('./Fornamn'),
        $namn_node->findvalue('./Efternamn'), "\n";
    
    # Using Namsokning and termdata format:
    
    use Navet::ePersondata::Namnsokning;
    
    my $ep = Navet::ePersondata::Namnsokning->new(
        pkcs12_file => '/path/to/certificate.p12',
        pkcs12_password => '5085873593180405',
        OrgNr => '162021004748',
        BestallningsId => '00000079-FO01-0001',
        format => 'termdata'
    );
    
    my $query = {
        Kon => 'K',
        Fornamn => 'sar*',
        Postort => 'Boden'            
    };
    
    foreach my $termdata ( $ep->find_all($query) ) {
        my ($first_name, $middle_name, $last_name) =
            ($termdata->{'01012'}, $termdata->{'01013'}, $termdata->{'01014'});
            
        $full_name = $first_name
            . ($middle_name ? " $middle_name" : '') . " $last_name";
        print "$full_name\n";
    }
  

=head1 DESCRIPTION

This package is the base class for talking with  Web Services - ePersondata
provided by skatteverket.se. The specific modules to use are
L<Navet::ePersondata::PersonPost> and  L<Navet::ePersondata:Namsokning> corresponding
to Web Service Personpost and Namnsökning. Using these services requires agreement
and access granted by Skatteverket in Sweden and also a special certificate. 

=head1 METHODS

=head2 new( %options )

    my $personpost_client = Navet::ePersondata::Personpost->new(
       pkcs12_file => '/path/to/certificate_file',
       pkcs12_password  => 'secret', 
       OrgNr => '162021004748',  
       BestallningsId => '00000079-FO01-0001',
       format => 'termdata', # ('termdata' or 'xml') default 'xml'
       soap_options => { proxy => 'https..', ...}  
    );

This constructor requires C<pkcs12_file>, C<pkcs12_password>, C<OrgNr> and
C<BestallningsId> or it will croak.

See information from Skatteverket in how to optain these. See
L<SOAP::XML::Client> for C<soap_options> if they are needed.

=head2 find_all( \%query_data )

    my @hits = $personpost_client->find_all( {PersonId => '196601010101'} )
    
query_data should contain the query parameters to the service.

Performs a SOAP call to the service and returns a list of objects that match the query.
The format of the objects is determined by what format was specified in the constructor:
L<XML::LibXML::Node> if it was 'xml', references to a flat hash of termcodes, termdata
if it was 'termdata'. Returns empty list if there was an error or if no record was found.

=head2 find_first( \%query_data )
   
Like find_all except returns a single entry (first found) instead of a list.
If not found or an error occurred it returns C<undef>.

=head2 error()

    if (my $err = $client->error) {
        die "Error: message=", $err->{message}, "\n";
    }

Returns a reference to a hash of the error information if the last call resulted in an error.
Returns C<undef> if there was no error. The hash can contain the following fields:

    message          - error text
    soap_faultcode   - SOAP faultcode from /Envelope/Body/Fault/faultscode
    soap_faultstring - SOAP faultstring from /Envelope/BodyFault/faultstring
    sv_Felkod        - Extra error code provided by Skatteverket
    sv_Beskrivning   - Extra description provided by Skatteverket
    raw_error        - Unparsed error text (can be XML, HTML or plain text)
    https_status     - HTTP status code

Only C<message> and C<raw_error> are garanteed to be set.

=head1 SEE ALSO

L<Navet::ePersondata::Personpost>

L<Navet::ePersondata::Namnsokning>

L<SOAP::XML::Client>

L<https://skatteverket.se/foretagochorganisationer/myndigheter/informationsutbytemellanmyndigheter/navethamtauppgifteromfolkbokforing.html>  

=head1 AUTHOR

Snorri Briem E<lt>snorri.briem@ub.lu.seE<gt>

=head1 COPYRIGHT

Copyright 2018- Lund University Library

=head1 LICENSE

This is free software; you can redistribute it and/or modify it under the same terms as the Perl 5 programming language system itself.

=cut
